### ISSUE TITLE
HTTP2: Server doesn't completely read request - The CancellationTokenSource has been disposed.

### ISSUE THREAD

**Issue by JamesNK** [dotnet/corefx#40207]
I have created a repro for an error that showed up on the CI build.

Repro:
1. `git clone https://github.com/JamesNK/grpc-dotnet.git`
2. `git checkout jamesnk/unimplemented-parallel`
3. `dotnet test test\FunctionalTests --filter Name~DuplexStream_SendToUnimplementedMethodAfterResponseReceived_MoveNextThrowsError > test.log` (writing to a file because it outputs a lot of logging!)

Logic in the test is something like:
1. Client makes a duplex streaming call to the server. The server will immediately finish the call without reading the request body
2. Client waits for response headers (this is where the HttpClient error eventually happens)
3. Client closes request stream (SerializeToStreamAsync completes)
4. Client attempts to read from the response stream
5. Go to 1.

I don't know if 3 & 4 are important. They just happen to be part of the test.

When one thread is executing this **sometimes** errors.

When multiple threads share one HttpClient it consistently throws this error:

```
System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.ObjectDisposedException: The CancellationTokenSource has been disposed.
   at System.Threading.CancellationTokenSource.ThrowObjectDisposedException()
   at System.Threading.CancellationTokenSource.Cancel()
   at System.Net.Http.Http2Connection.Http2Stream.OnReset(Exception resetException, Nullable`1 resetStreamErrorCode, Boolean canRetry)
   at System.Net.Http.Http2Connection.ProcessRstStreamFrame(FrameHeader frameHeader)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage request) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 565
   at Grpc.Net.Client.Internal.GrpcCall`2.GetResponseHeadersAsync() in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 266
   at Grpc.Tests.Shared.TaskExtensions.TimeoutAfter[T](Task`1 task, TimeSpan timeout, String filePath, Int32 lineNumber) in C:\Development\Source\grpc-dotnet\test\Shared\TaskExtensions.cs:line 59
   at Grpc.AspNetCore.FunctionalTests.Client.StreamingTests.<>c__DisplayClass3_0.<<DuplexStream_SendToUnimplementedMethodAfterResponseReceived_MoveNextThrowsError>b__1>d.MoveNext() in C:\Development\Source\grpc-dotnet\test\FunctionalTests\Client\StreamingTests.cs:line 197
```

```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview9-013753
 Commit:    3a441e719f
```

// @geoffkizer

**Comment by JamesNK** [comment]
I think this error was hidden behind https://github.com/dotnet/corefx/pull/40112. Now that lots of "incomplete" calls are possible this has shown up when the client is stressed.

**Comment by davidsh** [comment]
@stephentoub

**Comment by geoffkizer** [comment]
I think we can probably just not dispose the CTS. @stephentoub any risk from doing this?

**Comment by stephentoub** [comment]
> I think we can probably just not dispose the CTS. @stephentoub any risk from doing this?

As long as we're creating it with `new CancellationTokenSource()`, it should be ok. We should add a big comment on the creation site: if we instead create it would CreateLinkedTokenSource, not disposing it could create a leak. We would also want to avoid using `CancelAfter`, as that could create a Timer that we'd then leave hanging around until it fired.

**Comment by stephentoub** [comment]
> I have created a repro for an error that showed up on the CI build.

I'm trying to get the repro to fail, but I've run it several times now and the log shows results like the following each time:
```
Test run for d:\repos\grpc-dotnet\test\FunctionalTests\bin\Debug\netcoreapp3.0\Grpc.AspNetCore.FunctionalTests.dll(.NETCoreApp,Version=v3.0)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0-preview-20190808-04
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.

Test Run Successful.
Total tests: 1
     Passed: 1
 Total time: 48.9935 Seconds
```
Is there a tweak I need to make to it to get it to fail?

```
.NET Core SDKs installed:
  3.0.100-preview9-013795 [C:\Program Files\dotnet\sdk]
```

**Comment by stephentoub** [comment]
Reopening to track release/3.0 port.

**Comment by davidsh** [comment]
Fixed in release/3.0 branch with PR #40264
